source/fuzz/transformation_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,164 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_wrap_vector_synonym.h"
+
+#include "source/fuzz/data_descriptor.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    protobufs::TransformationWrapVectorSynonym message)
+    : message_(std::move(message)) {}
+
+TransformationWrapVectorSynonym::TransformationWrapVectorSynonym(
+    uint32_t instruction_id, uint32_t vector_operand1, uint32_t vector_operand2,
+    uint32_t fresh_id, uint32_t pos) {
+  message_.set_instruction_id(instruction_id);
+  message_.set_vector_operand1(vector_operand1);
+  message_.set_vector_operand2(vector_operand2);
+  message_.set_fresh_id(fresh_id);
+  message_.set_scalar_position(pos);
+}
+
+bool TransformationWrapVectorSynonym::IsApplicable(
+    opt::IRContext* ir_context,
+    const TransformationContext& transformation_context) const {
+  // |fresh_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  auto instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.instruction_id());
+
+  // |instruction_id| must refer to an existing instruction.
+  if (instruction == nullptr) {
+    return false;
+  }
+
+  auto type_instruction =
+      ir_context->get_def_use_mgr()->GetDef(instruction->type_id());
+
+  // The instruction must be of a valid scalar operation type.
+  if (!OpcodeIsSupported(instruction->opcode()) ||
+      !OperandTypeIsSupported(type_instruction)) {
+    return false;
+  }
+
+  assert(!transformation_context.GetFactManager()->IdIsIrrelevant(
+             instruction->result_id()) &&
+         "Result id of the scalar operation must be relevant.");
+
+  // |vector_operand1| and |vector_operand2| must exist.
+  auto vec1 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand1());
+  auto vec2 = ir_context->get_def_use_mgr()->GetDef(message_.vector_operand2());
+
+  if (vec1 == nullptr || vec2 == nullptr) {
+    return false;
+  }
+
+  // The 2 vectors must be the same valid vector type.
+  auto vec1_type_id = vec1->type_id();
+  auto vec2_type_id = vec2->type_id();
+
+  if (vec1_type_id != vec2_type_id) {
+    return false;
+  }
+
+  if (ir_context->get_def_use_mgr()->GetDef(vec1_type_id)->opcode() !=
+      SpvOpTypeVector) {
+    return false;
+  }
+
+  // |scalar_position| needs to be a non-negative integer less than the vector
+  // length.
+  // OpTypeVector instruction has the component count at index 2.
+  if (message_.scalar_position() >= ir_context->get_def_use_mgr()
+                                        ->GetDef(vec1_type_id)
+                                        ->GetSingleWordInOperand(0)) {
```

Comment:
Should this be `<` instead of `>=`? Also, I think you need to do `...->GetSingleWordInOperand(1)` to get the number of components in a vector.

Reply:
We want to return false on this in case it is out of bound of [0, |component_count], so I think it should be larger than or equal to.

Reply:
Right, my bad :) The component index is incorrect, though :)

Reply:
I'm testing on it right now just to make sure, but I think you are right. My test actually was bad on the position coverage as well.

Reply:
Done.